File: ehmp_1.2.0/ehmp_1.2.0/ehmp_1.2.0/ehmp/product/production/vista-rpc-resource/src/main/java/com/vistacowboy/jVista/VistaSelect.java
package com.vistacowboy.jVista;

import java.util.LinkedHashMap;

/**
 * Created with IntelliJ IDEA.
 * User: Joe
 * Date: 11/27/12
 * Time: 10:08 AM
 */

public class VistaSelect
{
    private String file;
    private String iens;
    private String fields = "@";
    private String flags = "IP";
    private int number = -1;
    private String from;
    private String part;
    private String index = "#";
    private String screen;
    private String identifier;
    private String[][] records;

    public VistaSelect() { }

    public String getFile()
    {
        return file;
    }

    public void setFile(String value)
    {
        file = value;
    }

    public String getIens()
    {
        return iens;
    }

    public void setIens(String value) throws VistaException
    {
        if (!value.startsWith(","))
        {
            value = ',' + value;
        }
        if (!value.endsWith(","))
        {
            value += ',';
        }
        String[] parts = value.substring(1, value.length()-1).split(",", -1);
        for (String ien : parts)
        {
            if (!VistaUtils.isNumeric(ien))
            {
                throw new VistaException(String.format("Non-numeric IEN: %s", ien));
            }
        }
        iens = value;
    }

    public String getFields()
    {
        return fields;
    }

    public void setFields(String value)
    {
        if (!value.isEmpty())
        {
            fields = value.contains("@") ? value : fields + ';' + value;
        }
    }

    public String getFlags()
    {
        return flags;
    }

    public void setFlags(String value) throws VistaException
    {
        if (value.isEmpty())
        {
            value = "IP";
        }
        else if (!value.contains("P"))
        {
            throw new VistaException("Current version does packed queries only");
        }
        flags = value;
    }

    public String getIndex()
    {
        return index;
    }

    public void setIndex(String value)
    {
        index = value;
    }

    public int getNumber()
    {
        return number;
    }

    public void setNumber(int value)
    {
        number = value;
    }

    public String getFrom()
    {
        return from;
    }

    public void setFrom(String value)
    {
        from = value;
    }

    public String getPart()
    {
        return part;
    }

    public void setPart(String value)
    {
        part = value;
    }

    public String getScreen()
    {
        return screen;
    }

    public void setScreen(String value)
    {
        screen = value;
    }

    public String getIdentifier()
    {
        return identifier;
    }

    public void setIdentifier(String value)
    {
        identifier = value;
    }

    public String[][] getRecords()
    {
        return records;
    }

    public String[][] find(VistaConnection cxn) throws VistaException
    {
        String rpc = prepare();
        String response = cxn.exec(rpc);
        load(response);
        return records;
    }

    private String prepare() throws VistaException
    {
        LinkedHashMap<String, String> param_list = prepareParamList();
        RpcParameter param = new RpcParameter(RpcParameter.LIST, param_list);
        return VistaRpc.prepare("DDR LISTER", new RpcParameter[]{param});
    }

    private LinkedHashMap<String, String> prepareParamList() throws VistaException
    {
        if (file == null || file.isEmpty())
        {
            throw new VistaException("VistaSelect must specify a file");
        }
        LinkedHashMap<String, String> param_list = new LinkedHashMap<String, String>();
        param_list.put("\"FILE\"", file);

        if (iens != null && !iens.isEmpty())
        {
            param_list.put("\"IENS\"", iens);
        }

        param_list.put("\"FIELDS\"", fields);
        param_list.put("\"FLAGS\"", flags);

        if (number != -1)
        {
            param_list.put("\"MAX\"", String.valueOf(number));
        }

        if (from != null && !from.isEmpty())
        {
            param_list.put("\"FROM\"", VistaUtils.adjustForSearch(from));
        }

        if (part != null && !part.isEmpty())
        {
            param_list.put("\"PART\"", part);
        }

        param_list.put("\"XREF\"", index);

        if (screen != null && !screen.isEmpty())
        {
            param_list.put("\"SCREEN\"", screen);
        }

        if (identifier != null && !identifier.isEmpty())
        {
            param_list.put("\"ID\"", identifier);
        }

        return param_list;
    }

    private void load(String response) throws VistaException
    {
        String[] lines = response.split("\r\n");

        // Find the starting line...
        int linenum = 0;
        while (linenum < lines.length && !lines[linenum].equals("[BEGIN_diDATA]") && !lines[linenum].equals("[BEGIN_diERRORS]")) linenum++;
        if (linenum == lines.length)
        {
            throw new VistaException("Empty response");
        }
        linenum++;

        int nrex = lines.length - linenum - 1;
        String[] flds = lines[linenum].split("\\^");
        records = new String[nrex][flds.length];
        for (int recnum = 0; recnum < nrex; recnum++, linenum++)
        {
            records[recnum] = lines[linenum].split("\\^");
        }

        if (!lines[linenum].equals("[END_diDATA]") && !lines[linenum].equals("[END_diERRORS]"))
        {
            throw new VistaException("Response error: missing END_diDATA?");
        }

        if (lines[linenum].equals("[END_diERRORS]"))
        {
            throw new VistaException("Errors returned in response", records);
        }
    }
}

File: ehmp_1.2.0/ehmp_1.2.0/ehmp_1.2.0/ehmp/product/tests/vista-rpc-tests/rbVista/vista_select.rb
require 'extlib'
require_relative './vista_exception'
require_relative './vista_utils'
require_relative './rpc_parameter'
require_relative './vista_rpc'

class VistaSelect

  attr_accessor :file, :number, :from, :part, :index, :screen, :identifier
  attr_reader :iens, :fields, :flags, :rpc, :param_list, :response, :records

  def initialize
    @fields = '@'
    @flags = 'IP'
    @index = '#'
  end

  def find cxn
    prepare
    @response = cxn.exec @rpc
    load @response
  end

  def prepare
    prepare_param_list
    params = RpcParameter.new(RpcParameter::LIST, @param_list)
    @rpc = VistaRpc.prepare('DDR LISTER', [params])
  end
  private :prepare

  def iens=(iens)
    iens = ',' + iens unless iens[0] == ','
    iens += ',' unless iens[iens.length-1] == ','
    parts = iens[1..iens.length-2].split(',', -1)
    parts.each do |part|
      raise VistaException.new 'Invalid IENS' unless VistaUtils.is_integer?(part)
    end
    @iens = iens
  end

  def fields=(fieldstr)
    fieldstr = '@;' + fieldstr unless fieldstr[0] == '@'
    fieldstr.chop! if fieldstr[fieldstr.length-1] == ';'
    @fields = fieldstr
  end

  def flags=(flags)
    if flags.nil? || flags.empty?
      flags = 'IP'
    elsif flags.index('P').nil?
      raise VistaException.new 'Current version does packed queries only'
    end
    @flags = flags
  end

  def prepare_param_list
    raise VistaException.new 'DDR call must specify a file' if @file.nil?

    @param_list = Hash.new()
    @param_list['"FILE"'] = @file.to_s
    @param_list['"IENS"'] = @iens unless @iens.nil?
    @param_list['"FIELDS"'] = @fields
    @param_list['"FLAGS"'] = @flags
    @param_list['"MAX"'] = @number.to_s unless @number.nil?
    @param_list['"FROM"'] = VistaUtils.adjust_for_search(@from) unless @from.nil?
    @param_list['"PART"'] = @part unless @part.nil?
    @param_list['"XREF"'] =  @index.nil? ? '#' : @index
    @param_list['"SCREEN"'] = @screen unless @screen.nil?
    @param_list['"ID"'] = @identifier unless @identifier.nil?
  end
  private :prepare_param_list

  def load(response)
    lines = response.split("\r\n")
    numlines = lines.length

    # Find starting line...
    linenum = 0
    while linenum < numlines && lines[linenum] != '[BEGIN_diDATA]'
      linenum += 1
    end
    raise VistaException.new 'Empty response' if linenum == numlines
    linenum += 1

    @records = Array.new
    while linenum < numlines && lines[linenum] != '[END_diDATA]'
      @records.push lines[linenum].split('^')
      linenum += 1
    end
    @records
  end
  private :load

end